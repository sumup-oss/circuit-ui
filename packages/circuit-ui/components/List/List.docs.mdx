import { Status, Props, Story } from '../../../../.storybook/components';

# List

<Status.Stable />

The List component is used to group pieces of content—called "list items"—together.

<Story id="typography-list--base" />
<Props />

## Component variations

### Variants

The List component comes in two variants, `ordered` or `unordered`.

- Use an ordered list to imply sequence or order (e.g. when providing a set of instructions).
- Use an unordered list when the order of the items is meaningless.

<Story id="typography-list--variants" />

### Sizes

The List component has two available sizes, `one` and `two`, matching the [`Body` component](Typography/Body)'s sizes.

For consistency, the list's size should be the same as the surrounding content's size.

<Story id="typography-list--sizes" />

### Nested lists

Lists can be nested to display a hierarchy, e.g. in a table of contents. The nested list will be visually indented.

Instead of relying on nested lists, consider splitting the content into separate sections, labeled by headings (refer to the Accessibility section below).

<Story id="typography-list--nested" />

## Accessibility

Lists are fundamental elements for structuring content. They can group pieces of content, give an overview of them, and make parsing and accessing them easier.

### Best practices

#### Keep list items concise

Keep the text in each list item concise. If you need to display more (structured) content, consider splitting it into sections instead, labeled by headings.

#### Use tables for two-dimensional data

In order to display two-dimensional data, consider using a table instead of a list. This makes parsing easier for sighted users and allows screen reader users to move vertically and horizontally across the data.

So instead of structuring content inside a list item:

```html
<ul>
  <li>Germany (Capital: Berlin)</li>
  <li>Ireland (Capital: Dublin)</li>
  <li>Canada (Capital: Ottawa)</li>
</ul>
```

Consider displaying the data in a table (you can use the [`Table` component](Components/Table)):

```html
<table>
  <tr>
    <th>Country</th>
    <th>Capital</th>
  </tr>
  <tr>
    <td>Germany</td>
    <td>Berlin</td>
  </tr>
  <tr>
    <td>Ireland</td>
    <td>Dublin</td>
  </tr>
  <tr>
    <td>Canada</td>
    <td>Ottawa</td>
  </tr>
</table>
```

See also the section on definition lists below.

#### Description lists

- not supported by this component
- not recommended (except for term/definition key-pair values?): screen reader discrepancies

#### Using lists for non-text UI elements

- for lists of cards, e.g. a list of articles
- not everything should be a list (can be quite noisy for SR users). think of how it is announced to screen readers and whether it makes sense for the list to be announced

### Resources

#### Related WCAG success criteria

- 1.3.1: [Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
