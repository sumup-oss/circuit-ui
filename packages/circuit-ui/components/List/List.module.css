.base {
  font-weight: var(--cui-font-weight-regular);
}

/* Sizes */

.one {
  padding-left: var(--cui-spacings-kilo);
  font-size: var(--cui-typography-body-m-font-size);
  line-height: var(--cui-typography-body-m-line-height);
}

.one li {
  margin-bottom: var(--cui-spacings-byte);
  margin-left: var(--cui-spacings-kilo);
}

.one li:last-child {
  margin-bottom: 0;
}

.one ul,
.one ol {
  margin-bottom: var(--cui-spacings-byte);
  margin-left: var(--cui-spacings-kilo);
}

.one ul:last-child,
.one ol:last-child {
  margin-bottom: 0;
}

.two {
  padding-left: var(--cui-spacings-kilo);
  font-size: var(--cui-typography-body-s-font-size);
  line-height: var(--cui-typography-body-s-line-height);
}

.two li {
  margin-bottom: var(--cui-spacings-byte);
  margin-left: var(--cui-spacings-bit);
}

.two li:last-child {
  margin-bottom: 0;
}

.two ul,
.two ol {
  margin-bottom: var(--cui-spacings-byte);
  margin-left: var(--cui-spacings-bit);
}

.two ul:last-child,
.two ol:last-child {
  margin-bottom: 0;
}
