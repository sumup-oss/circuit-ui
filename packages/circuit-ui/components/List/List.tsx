/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { forwardRef, Ref, OlHTMLAttributes } from 'react';
import { css } from '@emotion/react';
import isPropValid from '@emotion/is-prop-valid';

import styled, { StyleProps } from '../../styles/styled';
import { typography } from '../../styles/style-mixins';

type Size = 'one' | 'two';
type Variant = 'ordered' | 'unordered';

export interface ListProps extends OlHTMLAttributes<HTMLOListElement> {
  /**
   * A Circuit UI Body size. Should match surrounding text.
   */
  size?: Size;
  /**
   * Whether the list should be presented as an ordered or unordered list. Defaults to `unordered`.
   */
  variant?: Variant;
  /**
   The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLOListElement & HTMLUListElement>;
}

const baseStyles = ({ theme }: StyleProps) => css`
  font-weight: ${theme.fontWeight.regular};
`;

const sizeStyles = ({ theme, size = 'one' }: ListProps & StyleProps) => {
  const sizeMap = {
    one: {
      marginBottom: theme.spacings.byte,
      paddingLeft: theme.spacings.kilo,
      marginLeft: theme.spacings.kilo,
      type: typography('one')(theme),
    },
    two: {
      marginBottom: theme.spacings.kilo,
      paddingLeft: theme.spacings.kilo,
      marginLeft: theme.spacings.bit,
      type: typography('two')(theme),
    },
  };
  const { marginBottom, paddingLeft, marginLeft, type } = sizeMap[size];
  return css`
    padding-left: ${paddingLeft};
    ${type};

    li {
      margin-bottom: ${marginBottom};
      margin-left: ${marginLeft};
      &:last-child {
        margin-bottom: 0;
      }
    }

    ul,
    ol {
      margin-bottom: ${marginBottom};
      margin-left: ${marginLeft};
      &:last-child {
        margin-bottom: 0;
      }
    }
  `;
};

const BaseList = styled('ul', {
  shouldForwardProp: (prop) => isPropValid(prop) && prop !== 'size',
})<ListProps>(baseStyles, sizeStyles);

/**
 * A list, which can be ordered or unordered.
 */
export const List = forwardRef(
  ({ variant, ...props }: ListProps, ref: ListProps['ref']) => (
    <BaseList as={variant === 'ordered' ? 'ol' : 'ul'} {...props} ref={ref} />
  ),
);

List.displayName = 'List';
