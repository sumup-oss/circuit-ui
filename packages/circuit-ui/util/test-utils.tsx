/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
