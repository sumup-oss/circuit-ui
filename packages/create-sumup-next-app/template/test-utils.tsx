import React, { FunctionComponent } from 'react';
import { configureAxe } from 'jest-axe';
import { ThemeProvider } from '@emotion/react';
import { render as renderTest } from '@testing-library/react';
import { light } from '@sumup/design-tokens';

const WithProviders: FunctionComponent = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);

export const render = (
  component: Parameters<typeof renderTest>[0],
  options: Parameters<typeof renderTest>[1] = {},
) => renderTest(component, { wrapper: WithProviders, ...options });

export const axe = configureAxe({
  rules: {
    // Disable landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export * from '@testing-library/react';
