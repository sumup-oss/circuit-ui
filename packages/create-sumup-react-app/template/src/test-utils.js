import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';
import PropTypes from 'prop-types';

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);
AllTheProviders.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
