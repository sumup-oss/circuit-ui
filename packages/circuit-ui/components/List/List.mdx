import { Meta, Status, Props, Story } from '../../../../.storybook/components';
import * as Stories from './List.stories';

<Meta of={Stories} />

# List

<Status.Stable />

The List component is used to group pieces of content—the list items—together.

<Story of={Stories.Base} />
<Props />

## Component variations

### Variants

The List component comes in two variants, `ordered` or `unordered`.

- Use an ordered list to imply sequence or order (e.g. when providing a set of instructions).
- Use an unordered list when the order of the items is meaningless.

<Story of={Stories.Variants} />

### Sizes

The List component has two available sizes, `one` and `two`, matching the [`Body` component](Typography/Body)'s sizes.

For consistency, the list's size should be the same as the surrounding content's size.

<Story of={Stories.Sizes} />

### Nested lists

Lists can be nested to display a hierarchy, e.g. in a table of contents. The nested list will be visually indented.

Instead of relying on nested lists, consider splitting the content into separate sections, labeled by headings (refer to the Accessibility section below).

<Story of={Stories.Nested} />

---

## Accessibility

Lists are fundamental elements for content structure. They group related pieces of content, and make parsing and accessing them easier.

### Best practices

#### Keep list items concise

Keep the text in each list item concise. If you need to display more (structured) content, consider splitting it into sections instead, labeled by headings.

#### Use tables for two-dimensional data

In order to display two-dimensional data, consider using a table instead of a list. This makes parsing easier for sighted users, and allows screen reader users to move vertically and horizontally across the data.

For example, instead of structuring content inside a list item:

```html
<ul>
  <li>Germany (Capital: Berlin)</li>
  <li>Ireland (Capital: Dublin)</li>
  <li>Canada (Capital: Ottawa)</li>
</ul>
```

Consider displaying the data in a table (you can use the [`Table` component](Components/Table)):

```html
<table>
  <tr>
    <th>Country</th>
    <th>Capital</th>
  </tr>
  <tr>
    <td>Germany</td>
    <td>Berlin</td>
  </tr>
  <tr>
    <td>Ireland</td>
    <td>Dublin</td>
  </tr>
  <tr>
    <td>Canada</td>
    <td>Ottawa</td>
  </tr>
</table>
```

#### Description lists

##### What are descriptions lists?

One of the most obscure HTML elements, [description lists](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl) (`dl`) were initially intended for displaying glossary-like name-value pairs: a term (`dt`) and its definition (`dd`). Newer versions of the HTML spec also allow other name-value pairs, such as metadata labels and values, or questions and answers:

```html
<dl>
  <!-- term and definition -->
  <dt>Circuit</dt>
  <dd>A closed, usually circular line that goes around an object or area.</dd>
  <!-- metadata label and value -->
  <dt>Account number</dt>
  <dd>12345</dd>
  <!-- question and answer -->
  <dt>How do I find my account number?</dt>
  <dd>You can find your account number under "Profile details".</dd>
</dl>
```

##### Avoid using description lists

Description lists are intentionally not supported by the List component. They are not consistently announced by screen readers, and can confuse users.

For example, consider this simple description list, containing 2 metadata label-value pairs:

<details>
  <summary>Show the rendered description list</summary>
  <dl>
    <dt>Account number</dt>
    <dd>12345</dd>
    <dt>Sign-up date</dt>
    <dd>21.12.2021</dd>
  </dl>
</details>

```html
<dl>
  <dt>Account number</dt>
  <dd>12345</dd>
  <dt>Sign-up date</dt>
  <dd>21.12.2021</dd>
</dl>
```

Here's the screen reader output on three major [screen reader / browser combinations](https://webaim.org/projects/screenreadersurvey9/#browsercombos), when arrowing through the list:

| OS         | Browser | Screen reader | Output                                                                                                     |
| ---------- | ------- | ------------- | ---------------------------------------------------------------------------------------------------------- |
| Windows 10 | Chrome  | JAWS          | Definition list of 2 items. Account number. 12345. Sign-up date. 21.12.2021.                               |
| Windows 10 | Chrome  | NVDA          | List with 4 items. Account number. 12345. Sign-up date. 21.12.2021.                                        |
| macOS      | Safari  | VoiceOver     | Description list 2 items. Account number, 1 of 4. 12345, 2 of 4. Sign-up date, 3 of 4. 21.12.2021, 4 of 4. |

As you can see, there are inconsistencies between outputs:

- The description list has 2 name-value pairs, but NVDA on Chrome announces it as a "list with 4 items"
- VoiceOver on macOS Safari correctly announces "2 items", but goes on to announce each item's position on a total of 4 items ("Account number, 1 of 4")
- JAWS on Chrome announces a "definition list" (the initial name of the `dl` element), although this description list doesn't contain definitions
- Other screen reader / browser combinations (such as NVDA on Firefox or VoiceOver on Chrome) have yet other behaviors

This can be very confusing to users, so we recommend against using description lists at this time. Instead, use an unordered list or a table.

#### Using lists for other UI elements

In some cases, a list shouldn't only have text as list items, but another UI element such as a card (e.g. for a list of articles on a news website).

The List component doesn't support this use case. Instead, use an HTML unordered list element:

```jsx
function ArticleList(articles) {
  return (
    <ul
      css={css`
        list-style: none;
      `}
      role="list"
    >
      {articles.map((article, i) => (
        <li key={i}>
          <ArticleCard article={article} />
        </li>
      ))}
    </ul>
  );
}
```

**Warning**: if you remove the list's markers using `list-style: none;`, you'll need to add `role="list"` to the `ul` element. This is because [VoiceOver on Webkit removes list semantics when list markers are hidden](https://www.scottohara.me/blog/2019/01/12/lists-and-safari.html), both on iOS and macOS.

Bear in mind that not every group of elements should necessarily be a list. Lists can be quite verbose for screen-reader users, and using them excessively can be counter-productive. Imagine your list being announced as a "list of n items" and each subsequent item as "item m of n", and whether this improves or hinders user experience.

### Resources

#### Related WCAG success criteria

- 1.3.1: [Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html)
