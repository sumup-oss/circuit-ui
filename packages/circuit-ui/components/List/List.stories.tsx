/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Fragment } from 'react';

import Body from '../Body';
import { spacing } from '../../styles/style-mixins';

import docs from './List.docs.mdx';
import { List, ListProps } from './List';

export default {
  title: 'Typography/List',
  component: List,
  parameters: {
    docs: { page: docs },
  },
};

const ListItems = () => (
  <Fragment>
    <li>This is a list</li>
    <li>A very fine list</li>
  </Fragment>
);

export const Base = (args: ListProps) => (
  <List {...args}>
    <ListItems />
  </List>
);

const variants: ListProps['variant'][] = ['unordered', 'ordered'];

export const Variants = (args: ListProps) =>
  variants.map((variant) => (
    <List
      key={variant}
      {...args}
      variant={variant}
      css={spacing({ bottom: 'giga' })}
    >
      <ListItems />
    </List>
  ));

const sizes: ListProps['size'][] = ['one', 'two'];

export const Sizes = (args: ListProps) =>
  sizes.map((size) => (
    <>
      <Body size={size} css={spacing({ bottom: 'mega' })}>
        Use List size {size} with Body {size} text.
      </Body>
      <List key={size} {...args} size={size} css={spacing({ bottom: 'giga' })}>
        <ListItems />
      </List>
    </>
  ));

export const Nested = (args: ListProps) => (
  <List {...args}>
    <li>
      This is a list
      <List {...args}>
        <li>Sometimes a nested list</li>
      </List>
    </li>
    <li>And the list goes on</li>
  </List>
);
