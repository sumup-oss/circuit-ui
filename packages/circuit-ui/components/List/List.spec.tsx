/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it } from 'vitest';
import { createRef } from 'react';

import { render, axe } from '../../util/test-utils.js';

import { List } from './List.js';

describe('List', () => {
  it('should merge a custom class name with the default ones', () => {
    const className = 'foo';
    const { container } = render(
      <List className={className}>
        <li>List</li>
      </List>,
    );
    const list = container.querySelector('ul');
    expect(list?.className).toContain(className);
  });

  it('should forward a ref', () => {
    const ref = createRef<HTMLOListElement>();
    const { container } = render(
      <List ref={ref}>
        <li>List</li>
      </List>,
    );
    const list = container.querySelector('ul');
    expect(ref.current).toBe(list);
  });

  it('should meet accessibility guidelines', async () => {
    const { container } = render(
      <List>
        <li>List</li>
      </List>,
    );
    const actual = await axe(container);
    expect(actual).toHaveNoViolations();
  });
});
