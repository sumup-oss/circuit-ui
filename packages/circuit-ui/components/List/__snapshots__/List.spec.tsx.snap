// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render a nested list 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 1rem;
  line-height: 1.5rem;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ul
  class="circuit-0"
>
  <li>
    First level
  </li>
  <ul
    class="circuit-0"
  >
    <li>
      Second level
    </li>
  </ul>
</ul>
`;

exports[`List should render the ordered variant 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 1rem;
  line-height: 1.5rem;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ol
  class="circuit-0"
>
  <li>
    ordered
  </li>
</ol>
`;

exports[`List should render the unordered variant 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 1rem;
  line-height: 1.5rem;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ul
  class="circuit-0"
>
  <li>
    unordered
  </li>
</ul>
`;

exports[`List should render with default styles 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 1rem;
  line-height: 1.5rem;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ul
  class="circuit-0"
>
  <li>
    List
  </li>
</ul>
`;

exports[`List should render with size one 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 1rem;
  line-height: 1.5rem;
}

.circuit-0 li {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 8px;
  margin-left: 12px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ul
  class="circuit-0"
>
  <li>
    one
  </li>
</ul>
`;

exports[`List should render with size two 1`] = `
.circuit-0 {
  font-weight: 400;
  padding-left: 12px;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.circuit-0 li {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 li:last-child {
  margin-bottom: 0;
}

.circuit-0 ul,
.circuit-0 ol {
  margin-bottom: 12px;
  margin-left: 4px;
}

.circuit-0 ul:last-child,
.circuit-0 ol:last-child {
  margin-bottom: 0;
}

<ul
  class="circuit-0"
>
  <li>
    two
  </li>
</ul>
`;
